File: leetcode/python/Binary Tree Level Order Traversal/Binary_Tree_Level_Order_Traversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param root, a tree node
    # @return a list of lists of integers
    def levelOrder(self, root):
        if root is None:
            return []
        queue = []
        level = []
        path = []
        queue.append(root)
        queue.append(None)
        while queue:
            root = queue.pop(0)
            if root:
                level.append(root.val)
                if root.left:
                    queue.append(root.left)
                if root.right:
                    queue.append(root.right)
            else:
                path.append(level[:])
                level = []
                if queue:
                    queue.append(None)
        return path

File: leetcode/cpp/BinaryTreeLevelOrderTraversal.cpp
//
// Created by zhangqi on 2018/1/31.
// Time:  O(n)
// Space: O(n)

#include <vector>
#include <queue>
#include "TreeNode.h"

using namespace std;

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {

        vector<vector<int>> result;
        queue<TreeNode *> que;
        if(root != nullptr){
            que.emplace(root);
        }
        while (!que.empty()){
            vector<int> level;
            int size = que.size();
            for(int i = 0; i < size; i++){
                TreeNode *front = que.front();
                level.push_back(front->val);
                que.pop();

                if(front->left){
                    que.emplace(front->left);
                }

                if(front->right){
                    que.emplace(front->right);
                }
            }
            result.push_back(level);
        }
        return result;
    }
};